File: 219.存在重复元素-ii.py
#
# @lc app=leetcode.cn id=219 lang=python
#
# [219] 存在重复元素 II
#
# 解法1(T78% S60%): 还是通过哈希表，value保存下标，如果满足绝对距离小于k则返回，否则更新key的value下标。 为什么可以不用数组保存所有下标呢？因为如果之前的绝对距离太大不满足条件，可以放心大胆的舍弃换成最近的

# @lc code=start
class Solution(object):
    def containsNearbyDuplicate(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: bool
        """
        table = {}
        for i in range(len(nums)):
            if nums[i] not in table: table[nums[i]] = i
            else:
                if abs(i-table[nums[i]]) <= k: return True
                else: table[nums[i]] = i
        return False
# @lc code=end

File: javascript/219.存在重复元素-ii.js
/*
 * @lc app=leetcode.cn id=219 lang=javascript
 *
 * [219] 存在重复元素 II
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {boolean}
 */
var containsNearbyDuplicate = function(nums, k) {
    var map = new Map();
    for (var i=0; i<nums.length; i++) {
        if (map.has(nums[i])) {
            map.set(nums[i], map.get(nums[i]).concat(i));
        } else {
            map.set(nums[i], [i]);
        }
    }

    for (var arr of map.values()) {
        if (arr.length > 1) {
            arr.sort((a,b)=>(a-b));
            for (var i=1; i<arr.length; i++) {
                if (arr[i]-arr[i-1] <= k) {
                    return true;
                }
            }
        }
    }
    return false;
};
// @lc code=end

